### PR TITLE
 Alertmanager: Implement experimental /templates/test API

### PR DESCRIPTION
**Pull request by fayzal-g**
#### What this PR does
This PR adds experimental functionality to test templates under `api/v1/grafana/templates/test`.

#### Checklist

- [ ] Tests updated.
- [ ] Documentation added.
- [ ] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`.
- [ ] [`about-versioning.md`](https://github.com/grafana/mimir/blob/main/docs/sources/mimir/configure/about-versioning.md) updated with experimental features.

### CODE REVIEW

File: integration/e2emimir/client.go [hunk 1]
```diff
@@ -1479,6 +1480,41 @@ func (c *Client) GetReceiversExperimental(ctx context.Context) ([]alertingmodels
...
+	}
+
+	if resp.StatusCode/100 != 2 {
+		return nil, fmt.Errorf("testing templates failed with status %d and error %v", resp.StatusCode, string(body))
+	}
+
+	decoded := alertingNotify.TestTemplatesResults{}
+	if err := json.Unmarshal(body, &decoded); err != nil {
+		return nil, err
+	}
```
Comment (santihernandezc) [inline integration/e2emimir/client.go:1513]: I'm getting JSON field names capitalized, we should add JSON struct tags in the Alerting repo for `TestTemplatesResults`.

File: pkg/alertmanager/alertmanager.go [hunk 2]
```diff
@@ -342,6 +345,33 @@ func (am *Alertmanager) GetReceiversHandler(w http.ResponseWriter, _ *http.Reque
...
+			http.StatusInternalServerError)
+	}
+
+	w.Header().Set("Content-Type", "application/json")
+	if err := json.NewEncoder(w).Encode(response); err != nil {
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+		return
+	}
+}
+
```
Comment (santihernandezc) [inline pkg/alertmanager/alertmanager.go:374]: We're not returning after calling `http.Error()`, so the function continues executing and writing to the response writer. HTTP response I got (summarized):

```bash
HTTP/1.1 400 Bad Request
Content-Type: application/json

error unmarshalling test templates config JSON: invalid character '}' after object key
{"Results":null,"Errors":null}
```

In any case, submitting a bad request to Grafana's current template testing endpoint results in a JSON response. Any reason why we're not following the same pattern here?

File: pkg/alertmanager/alertmanager.go [hunk 3]
```diff
@@ -342,6 +345,33 @@ func (am *Alertmanager) GetReceiversHandler(w http.ResponseWriter, _ *http.Reque
 	}
 }
 
+func (am *Alertmanager) TestTemplatesHandler(w http.ResponseWriter, r *http.Request) {
+	c := alertingNotify.TestTemplatesConfigBodyParams{}
+	if err := json.NewDecoder(r.Body).Decode(&c); err != nil {
+		http.Error(w,
+			fmt.Sprintf("error unmarshalling test templates config JSON: %s", err.Error()),
+			http.StatusBadRequest)
```
Comment (santihernandezc) [inline pkg/alertmanager/alertmanager.go:353]:
```suggestion
		http.Error(w,
			fmt.Sprintf("error unmarshalling test templates config JSON: %s", err.Error()),
			http.StatusBadRequest)
		return
```

File: pkg/alertmanager/alertmanager.go [hunk 4]
```diff
@@ -342,6 +345,33 @@ func (am *Alertmanager) GetReceiversHandler(w http.ResponseWriter, _ *http.Reque
...
+	am.templatesMtx.RLock()
+	tmpls := make([]alertingTemplates.TemplateDefinition, len(am.templates))
+	copy(tmpls, am.templates)
+	am.templatesMtx.RUnlock()
+
+	response, err := alertingNotify.TestTemplate(r.Context(), c, tmpls, am.cfg.ExternalURL.String(), am.logger)
+	if err != nil {
+		http.Error(w,
+			fmt.Sprintf("error testing templates: %s", err.Error()),
+			http.StatusInternalServerError)
```
Comment (santihernandezc) [inline pkg/alertmanager/alertmanager.go:365]:
```suggestion
		http.Error(w,
			fmt.Sprintf("error testing templates: %s", err.Error()),
			http.StatusInternalServerError)
		return
```

File: pkg/alertmanager/distributor.go [hunk 5]
```diff
@@ -75,7 +75,14 @@ func (d *Distributor) isQuorumWritePath(p string) bool {
...
 func (d *Distributor) isUnaryWritePath(p string) bool {
-	return strings.HasSuffix(p, "/silences")
+	if strings.HasSuffix(p, "/silences") {
+		return true
+	}
+	if strings.HasSuffix(p, "/api/v1/grafana/templates/test") {
+		return true
+	}
+
+	return false
```
Comment (santihernandezc) [inline pkg/alertmanager/distributor.go:85]: We can condense this into one line:

```suggestion
	return strings.HasSuffix(p, "/silences") || strings.HasSuffix(p, "/api/v1/grafana/templates/test")
```

**Review by santihernandezc — CHANGES_REQUESTED** [review]

**Comment by ying-jeanne** [comment]
Hi, thanks for the PR, do you also needs this to be vendored for template test? https://github.com/grafana/mimir/pull/8560/files, it would be nice if someone knows well alert gives this PR a review

**Review by santihernandezc — APPROVED** [review]
LGTM, nice job 😎

File: integration/e2emimir/client.go [hunk 6]
```diff
@@ -1479,6 +1480,41 @@ func (c *Client) GetReceiversExperimental(ctx context.Context) ([]alertingmodels
...
+	u := c.alertmanagerClient.URL("api/v1/grafana/templates/test", nil)
+
+	data, err := json.Marshal(ttConf)
+	if err != nil {
+		return nil, fmt.Errorf("error marshalling test templates config: %s", err)
+	}
+
+	req, err := http.NewRequest(http.MethodPost, u.String(), bytes.NewReader(data))
+	if err != nil {
+		return nil, fmt.Errorf("error creating request: %v", err)
```
Comment (aknuds1) [inline integration/e2emimir/client.go:1493]: [Nit] Might as well be consistent :)
```suggestion
		return nil, fmt.Errorf("error creating request: %s", err)
```

File: integration/e2emimir/client.go [hunk 7]
```diff
@@ -1479,6 +1480,41 @@ func (c *Client) GetReceiversExperimental(ctx context.Context) ([]alertingmodels
...
+	if err != nil {
+		return nil, err
+	}
+
+	if resp.StatusCode == http.StatusNotFound {
+		return nil, ErrNotFound
+	}
+
+	if resp.StatusCode/100 != 2 {
+		return nil, fmt.Errorf("testing templates failed with status %d and error %v", resp.StatusCode, string(body))
```
Comment (aknuds1) [inline integration/e2emimir/client.go:1507]: [Nit] Might as well be precise about the type:
```suggestion
		return nil, fmt.Errorf("testing templates failed with status %d and error %s", resp.StatusCode, string(body))
```

**Review by aknuds1 — APPROVED** [review]
LGTM modulo nits.

**Pull request merged**